Prevent large error messages when using middleware instrumentation

zetter opened the pull request:
### Motivation / Background

When you use `ActiveSupport::Notifications` to instrument `process_middleware.action_dispatch`, certain errors may cause a very large error message to be generated. This can exhaust resources and make the application unresponsive

### Detail

This problem happens if the object that caused the exception has direct or indirect reference back to the application middleware stack and the error's exception message includes a representation of the object that raised. For example, a `NoMethod` error on an instance of `ActionDispatch::Request::Session` causes this as this has a reference to `ActionDispatch::Session::CookieStore` which contains a reference to the middleware stack through `@app`.

The error message grows incredibly large because the instrumenter runs around each middleware and handles an error by saving the error message to the `@payload` instance variable. When the next middleware runs, the it error message will include the string of the previously generated error message. This causes the length of the error to more than double for each middleware in the stack.

This effected a real application that I work on, and caused an error to be generated that was an estimated **61,000,000 characters long**.

This change prevents this problem by stopping the `@payload` instance variable from being inspected. `@payload` contains the `:exception_object` and `:exception` keys that cause the error message to grow and grow. The change makes sure `@middleware` is still inspected to match current behavior as much as possible.

This is reproducible in Rails 6.1.7.6, 7.0.8 and 7.1.0. Note that this is no longer an issue for `NoMethodError` in Ruby 3.3.0 preview 1, [as the exception message no longer includes the printed representation of the object](https://github.com/ruby/ruby/pull/6950). I haven't investigated if other standard library errors trigger this behavior.

I originally reported this to the Rails security team as it could be use to perform denial of service attacks on applications, but we decided that this is unlikely because of the very specific error that needs to be raised to trigger this issue.

### Additional information

Here's a minimal reproduction of the issue:

```ruby
# frozen_string_literal: true
require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  gem 'rails', '7.0.8'
  gem 'puma'
end

require "rails"
require "action_controller/railtie"
require "action_view/railtie"

class App < Rails::Application
  config.root = __dir__
  config.consider_all_requests_local = true
  config.secret_key_base = 'i_am_a_secret'
  config.eager_load = false

  routes.append do
    root to: 'welcome#index'
  end
end

ActiveSupport::Notifications.subscribe('process_middleware.action_dispatch') do |x|
  # just subscribing is enough to trigger this issue
end

class WelcomeController < ActionController::Base
  def index
    session[:key] = 'initialize-session'
    session.non_existant_method
    render inline: 'never reached'
  end
end

App.initialize!

Rack::Server.start(app: App)

# see error message by visiting / on app
```

Here's the [error the example produces before](https://github.com/rails/rails/files/12871979/example_error_before_patch.txt) and [after](https://github.com/rails/rails/files/12871978/example_error_after_patch.txt) the change.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

zzak commented:
Thanks for the patch!

I think this probably could use a changelog, since it's changing the behavior of `process_middleware.action_dispatch` notifications.

zetter committed:
Prevent large error messages when using middleware instrumentation

When you use `ActiveSupport::Notifications` to instrument
`process_middleware.action_dispatch`, certain errors may cause a very large
error message to be generated. This can exhaust resources and make the
application unresponsive.

This only happens if the object that caused the exception has direct or indirect
reference back to the application middleware stack and the error's exception
message includes a representation of the object that raised. For example, a
`NoMethod` error `ActionDispatch::Request::Session` causes this as this has a
reference to `ActionDispatch::Session::CookieStore` which contains a reference
to the middleware stack through `@app`.

The error message grows incredibly large because the instrumenter runs around
each middleware and handles an error by saving the error message to the
`@payload` instance variable. When the next middleware runs, the it error
message will include the string of the previously generated error message. This
causes the length of the error to more than double for each middleware in the
stack.

This effected a real application that I work on, and caused an error to be
generated that was an estimated **61,000,000 characters long**.

This change prevents this problem by stopping the `@payload` instance variable
from being inspected. `@payload` contains the `:exception_object` and
`:exception` keys that cause the error message to grow and grow. The change
makes sure `@middleware` is still inspected to match current behavior as much as
possible.

This is reproducible in Rails 6.1.7.6, 7.0.8 and 7.1.0. Note that this is no
longer an issue for `NoMethodError` in Ruby 3.3.0 preview 1, as the exception
message no longer includes the printed representation of the object[1]. I
haven't investigated if other standard library errors trigger this behavior.

I originally reported this to the Rails security team as it could be use to
perform denial of service attacks on applications, but we decided that this is
unlikely because of the very specific error that needs to be raised to trigger
this issue.

[1] - https://github.com/ruby/ruby/pull/6950

zetter commented:
> Thanks for the patch!
> 
> I think this probably could use a changelog, since it's changing the behavior of `process_middleware.action_dispatch` notifications.

Thanks for taking a look. I've added a changelog entry.

jonathanhefner commented on actionpack/lib/action_dispatch/middleware/stack.rb:
```diff
@@ -66,6 +66,10 @@ def call(env)
         end
       end
+
+      def inspect # :nodoc:
+        "#<#{self.class.name}:#{'%#016x' % (object_id << 1)} @middleware=#{@middleware.inspect}>"
+      end
```
What are the drawbacks of just passing through to `@middleware.inspect`?  i.e.:

```suggestion
      def inspect # :nodoc:
        @middleware.inspect
      end
```

Are there use cases where we want to inspect the middleware stack and see all the `ActionDispatch::MiddlewareStack::InstrumentationProxy` instances?

_Sidenote:_ Since Ruby 2.7 (I think), `'%#016x' % (object_id << 1)` does not yield the object's address.  The result is something like "0x00000000002df0" instead of "0x00007ff41537a788".  I don't think the difference is particularly important, but 16 hex digits might be overkill.

zetter replied:
> What are the drawbacks of just passing through to @middleware.inspect?

This would hide the `InstrumentationProxy` in the middleware stack. This could make debugging issues with middleware confusing (for example, stack traces won't match up with the inspected middleware) so my preference would be not to hide it.

> Sidenote: Since Ruby 2.7 (I think), '%#016x' % (object_id << 1) does not yield the object's address. The result is something like "0x00000000002df0" instead of "0x00007ff41537a788". I don't think the difference is particularly important, but 16 hex digits might be overkill.

I wasn't sure what to do here so followed the pattern of other inspect methods I found in Rails.
